earth_enterprise/src/common/khRefCounter.h

```diff
@@ -175,12 +190,30 @@ class khRefCounterImpl : public ThreadPolicy::MutexHolder {
  private:
   mutable uint32 refcount_;
 
-  // private and unimplemented
-  // classes derived from khRefCounterImpl are meant to be shared, not copied
-  khRefCounterImpl(const khRefCounterImpl &);
-  khRefCounterImpl & operator=(const khRefCounterImpl&);
-
  protected:
+  // protected and implemented to do nothing
+  // classes derived from khRefCounterImpl can decide if they want to support
+  // copy/move operations on their state but this base class state should
+  // do nothing druing these operations as this state is tied to memory managment
+  // and not really object state.
+  khRefCounterImpl(const khRefCounterImpl &) : refcount_(1) {
+    // intentionally left empty.  The right thing to do is nothing.
+  }
+  khRefCounterImpl & operator=(const khRefCounterImpl&) {
+    // intentionally left empty.  The right thing to do is nothing.
+    return *this;
+  }
+
+#if GEE_HAS_MOVE
+  khRefCounterImpl(khRefCounterImpl &&) noexcept : refcount_(1) {
```

Comment:
Q: I'm not sure how this class is being used, but aren't you supposed to copy the reference count when you're moving an object?

Reply:
No.  And the reason is the reference count (i.e. the base classes state) has nothing to do with the derived class (or its state).  This is one of those odd cases where the base class actually should support a do nothing move/copy.  The job of the base class has nothing to do with of whatever the derived class is trying to implement.  Instead the base class has the job of managing the one and only memory block it is assigned when first constructed and nothing more and that assignment never changes.  So as  the object state of the derived class changes because of move/copy (or anything else) this base class doesn't care because that doesn't have any influence on weather or not that block of memory can be cleaned up or not (i.e. it doesn't effect the number of external pointers pointing to this memory block).  The memory is cleaned up only when external intrusive pointers have zero references to that memory.  The fact that the state of the memory changed doesn't at all change how many external pointers are referencing this memory block (i.e. the following operation doesn't at all change the reference count: `*p1 = *p2`).  The base class has a completely different mission from the derived class: memory management.  So it can't and shouldn't participate in any assignment/move operations that the derived class is doing as that doesn't effect the job of memory management (for this memory block).  This is probably why the C++ committee instead added the make_shared<> as it gives you most of the same benefits of intrusive shared objects without the weirdness of things like this.

Reply:
I see.  I think this class just needs its documentation to say it just provides storage for a reference counter, but does nothing, and the derived classes are expected to make any use of the counter.